README.md
# React Native Image Cache on File System with Progressive Loading

[![npm version](https://img.shields.io/npm/v/@georstat/react-native-image-cache.svg?style=for-the-badge)](https://www.npmjs.com/package/@georstat/react-native-image-cache)
[![npm downloads](https://img.shields.io/npm/dm/@georstat/react-native-image-cache.svg?style=for-the-badge)](https://www.npmjs.com/package/@georstat/react-native-image-cache)
[![npm](https://img.shields.io/npm/dt/@georstat/react-native-image-cache.svg?style=for-the-badge)](https://www.npmjs.com/package/@georstat/react-native-image-cache)
[![License: MIT](https://img.shields.io/badge/License-MIT-green.svg?style=for-the-badge)](https://opensource.org/licenses/MIT)
![GitHub package.json version](https://img.shields.io/github/package-json/v/georstat/react-native-image-cache?style=for-the-badge)
![Platform - Android and iOS](https://img.shields.io/badge/platform-Android%20%7C%20iOS-blue.svg?style=for-the-badge)

Inspired by:

- [wcandillon/react-native-expo-image-cache](https://github.com/wcandillon/react-native-expo-image-cache) (William Candillon)
- [WrathChaos/react-native-progressive-fast-image](https://github.com/WrathChaos/react-native-progressive-fast-image) (FreakyCoder)

## Features

- Cache **remote** images in file system with progressive loading
- Can be used with local images
- Uses [`react-native-file-access`](https://github.com/alpha0010/react-native-file-access) for file system access
- Written in `Typescript`
- Uses [`react-native-reanimated`](https://docs.swmansion.com/react-native-reanimated/docs/) for animations
- Retry on network failure
- Supports new Architecture, both `react-native-file-access` and `react-native-reanimated` are Fabric ready.

## <span style="color:red;font-size: 26px;font-weight:bold;">Caution:</span>

#### If you're not using [`react-native-reanimated`](https://github.com/software-mansion/react-native-reanimated) version 2.x.x then use version 1.x.x of this module

### If you're looking for an image library with thumbnails visit [@georstat/react-native-image-gallery](https://github.com/georstat/react-native-image-gallery)

## Preview

### iOS:

![@georstat:react-native-image-cache_iOS_demo](https://user-images.githubusercontent.com/717975/117460761-85fd8f00-af55-11eb-8804-69c008bfeb8d.gif)

### Android:

![@georstat:react-native-image-cache_Android_demo](https://user-images.githubusercontent.com/717975/117491773-4ba5e900-af79-11eb-8ae1-99604b47c55c.gif)

### Custom Loading Image Component: (see [example](https://github.com/georstat/react-native-image-cache/tree/main/example/src))

![@georstat:react-native-image-cache_custom_loading_image_component_demo](https://user-images.githubusercontent.com/717975/125193571-b4e5f900-e255-11eb-8164-3ca097ce6e74.gif)

## Installation

#### yarn:

```bash
yarn add @georstat/react-native-image-cache react-native-file-access react-native-reanimated
```

#### npm:

```bash
npm i @georstat/react-native-image-cache react-native-file-access react-native-reanimated
```

#### link native packages:

```bash
 cd ios
 pod install
```

### \*\*note!

For react native >= 0.65 use [`react-native-file-access`](https://github.com/alpha0010/react-native-file-access) >= 2.0.0

## Usage

#### Put this Global config on your app entry eg. `App.tsx` or `index.js` (**Required**):

**_Note_**: `retryDelay` and `maxRetries` are optional, they default to `0` which means retry logic is **disabled**

```tsx
import { CacheManager } from '@georstat/react-native-image-cache';
import { Dirs } from 'react-native-file-access';

CacheManager.config = {
  baseDir: `${Dirs.CacheDir}/images_cache/`,
  blurRadius: 15,
  cacheLimit: 0,
  maxRetries: 3 /* optional, if not provided defaults to 0 */,
  retryDelay: 3000 /* in milliseconds, optional, if not provided defaults to 0 */,
  sourceAnimationDuration: 1000,
  thumbnailAnimationDuration: 1000,
};
```

#### Custom cache key, useful for caching images from Amazon's S3 and similar services, add this (or whatever suits your needs) on `CacheManager.config`:

```tsx
CacheManager.config = {
  // ...
  getCustomCacheKey: (source: string) => {
    // Remove params from the URL for caching images (useful for caching images from Amazons S3 bucket and etc)
    let newCacheKey = source;
    if (source.includes('?')) {
      newCacheKey = source.substring(0, source.lastIndexOf('?'));
    }
    return newCacheKey;
  },
};
```

#### `cacheLimit` config: (auto pruning of cached files)

If `cacheLimit` is set to `0` (default value) then the cache will never be auto pruned. This setting accepts a number of Bytes eg. `1024 * 1024 * 256`(~256MB) and requires `react-native-file-access` >= 2.4.0, if you're using < 2.4.0 then leave the default value `0` (disabled).

Cache pruning flow:

1. Get all files from `baseDir`.
2. Sort them by `lastModified`, oldest first.
3. Get total size in Bytes by using `reduce` array method.
4. Check total size from step 3 and subtract `cacheLimit` value.
5. Run a while loop and keep deleting files so long as the current cache size (step 4) is larger than the size required to trigger cache pruning (`cacheLimit` value).
6. All steps above will run only if the image is not already cached, it runs after downloading a new image into the cache.

Pruning has been benchmarked on iOS simulator with a 5.7MB image and ~5.000 copies of it on cache without any issues. Please note that the pruning speed/performance might differ among devices. Use `cacheLimit` wisely and do not set a big value.

If you want to run your own tests on simulator then the cached images are stored in this location on a Mac:
`/Users/<your_name>/Library/Developer/CoreSimulator/Devices/<simulator_device_id>/ data/Containers/Data/Application/<application_id>/Library/Caches/images_cache`,
just copy and paste multiple images in there, there's no need to download them via the app.

#### Directory constants, choose wisely: ([react-native-file-access docs](https://github.com/alpha0010/react-native-file-access#directory-constants))

- `Dirs.CacheDir`
- `Dirs.DatabaseDir` (Android only)
- `Dirs.DocumentDir`
- `Dirs.LibraryDir` (iOS only)
- `Dirs.MainBundleDir`
- `Dirs.SDCardDir` (Android only)
  - Prefer `FileSystem.cpExternal()` when possible.

#### Component:

```tsx
import { CachedImage } from '@georstat/react-native-image-cache';

<CachedImage
  source="https://via.placeholder.com/3500x3500"
  style={{ height: 350, width: 150 }}
  thumbnailSource="https://via.placeholder.com/350x150"
/>;
```

#### Prefetch Image(s) and store them in cache:

Accepts 2 parameters:

| Parameter | Type              | Description                                                                            |
|-----------|-------------------|----------------------------------------------------------------------------------------|
| `image`   | `Array or String` | (Required) uri of remote image or array of remote uri strings                          |
| `options` | `Object`          | (Optional) custom options for the fetch image http request eg. `{headers:{}, body:{}}` |

```tsx
import { CacheManager } from '@georstat/react-native-image-cache';

const url = 'https://..../image.jpg';

const urls = [
  'https://..../image.jpg',
  'https://..../image2.jpg',
  'https://..../image3.jpg',
];

CacheManager.prefetch(url); // prefetch single image

CacheManager.prefetch(urls); // prefetch mutliple images
```

#### Prefetch Image and return blob/base64:

Accepts 2 parameters:

| Parameter | Type     | Description                                                                            |
|-----------|----------|----------------------------------------------------------------------------------------|
| `image`   | `String` | (Required) uri of remote image                                                         |
| `options` | `Object` | (Optional) custom options for the fetch image http request eg. `{headers:{}, body:{}}` |

```tsx
import { CacheManager } from '@georstat/react-native-image-cache';

const url = 'https://..../image.jpg';

CacheManager.prefetchBlob(url).then(response => console.log(response)); // response is the base64 string
```

#### Clear local cache:

```tsx
import { CacheManager } from '@georstat/react-native-image-cache';

await CacheManager.clearCache();
```

#### Remove single cache entry:

```tsx
const uri = 'https://.../example.jpg';

await CacheManager.removeCacheEntry(uri);
```

#### Get local cache size:

```tsx
await CacheManager.getCacheSize();
```

#### Check if image is cached:

```tsx
await CacheManager.isImageCached(uri);
```

## Props

#### `CachedImage` accepts the following props:

| Properties                       | PropType             | Description                                                                                                                                                                                                    |
|----------------------------------|----------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `source`                         | `String`             | (**Required**) Uri of image.                                                                                                                                                                                   |
| `sourceAnimationDuration`        | `Number`             | `source` image animation duration when loading, defaults to `1000`ms (overrides config)                                                                                                                        |
| `thumbnailSource`                | `String`             | (**Required**) Uri of the thumbnail image                                                                                                                                                                      |
| `thumbnailAnimationDuration`     | `Number`             | Animation duration for thumbnail, defaults to `1000`ms (overrides config)                                                                                                                                      |
| `blurRadius`                     | `Number`             | Amount of blur to apply on `thumbnailSource` image , defaults to `15` (overrides config)                                                                                                                       |
| `loadingImageComponent`          | `React.CompnentType` | Defaults to an image with the `loadingSource` prop                                                                                                                                                             |
| `noCache`                        | `Boolean`            | Do not cache the image, defaults to `false` which means always cache the image                                                                                                                                 |
| `maxAge`                         | `Number`             | Maximum age in hours to cache the image, defaults to `undefined` (infinite caching). **Auto pruning won't take into consideration this value, it will delete the image anyway if the `cacheLimit` is reached** |
| `loadingImageStyle`              | `Object`             | Style for loading image component. Works if you don't provide a `loadingImageComponent`                                                                                                                        |
| `imageStyle`                     | `Object`             | Image style, use it when loading local images via `require()`                                                                                                                                                  |
| `loadingSource`                  | `object`             | Source for loading Image component. Works if you don't provide `loadingImageComponent`                                                                                                                         |
| `onError`                        | `Func`               | Runs when there is an error loading the image from cache                                                                                                                                                       |
| `onLoad`                         | `Func`               | Invoked when load completes successfully                                                                                                                                                                       |
| `onLoadEnd`                      | `Func`               | Invoked when load either succeeds or fails                                                                                                                                                                     |
| `resizeMode`                     | `String`             | React native Image component [resizeMode](https://reactnative.dev/docs/image#resizemode) defaults to `contain`                                                                                                 |
| `testID`                         | `String`             | testID, useful for tests                                                                                                                                                                                       |
| `tintColor`                      | `String`             | tintColor of the source image                                                                                                                                                                                  |
| `style`                          | `Object`             | `source` AND `thumbnailSource` image style                                                                                                                                                                     |
| `options`                        | `Object`             | custom options for the fetch image http request eg. `{headers:{}, body:{}}`                                                                                                                                    |
| `accessibilityHint`              | `string`             | accessibility hint for `source` (optional)                                                                                                                                                                     |
| `accessibilityLabel`             | `string`             | accessibility label for `source` (optional)                                                                                                                                                                    |
| `accessibilityRole`              | `string`             | accessibility role for `source` (optional, defaults to `image`)                                                                                                                                                |
| `accessibilityHintThumbnail`     | `string`             | accessibility hint for `thumbnailSource` (optional)                                                                                                                                                            |
| `accessibilityLabelThumbnail`    | `string`             | accessibility label for `thumbnailSource` (optional)                                                                                                                                                           |
| `accessibilityRoleThumbnail`     | `string`             | accessibility role for `thumbnailSource` (optional, defaults to `image`)                                                                                                                                       |
| `accessibilityHintLoadingImage`  | `string`             | accessibility hint for `loadingSource` (optional)                                                                                                                                                              |
| `accessibilityLabelLoadingImage` | `string`             | accessibility label for `loadingSource` (optional)                                                                                                                                                             |
| `accessibilityRoleLoadingImage`  | `string`             | accessibility role for `loadingSource` (optional, defaults to `image`)                                                                                                                                         |

#### [More info about React Native Accessibility](https://reactnative.dev/docs/accessibility)

#### [Blog Article about React Native Accessibility](https://medium.com/reactbrasil/introducing-react-native-accessibility-engine-fcf78f2a3805)

## Todo:

- [x] Convert library to React Hooks
- [x] Make `baseDir` configurable
- [x] Delete single cache entry

# Sponsor

This library is developed for free.

Buy us a coffee using this link: [!["Buy Me A Coffee"](https://www.buymeacoffee.com/assets/img/custom_images/orange_img.png)](https://www.buymeacoffee.com/efstathios)

## Authors:

- [Efstathios Ntonas](https://github.com/efstathiosntonas)
- [George Bakogiannis](https://github.com/geobako)
- [George Kallinikos](https://github.com/giokallis)


CHANGELOG.md
# Changelog

## [v3.1.0](https://github.com/georstat/react-native-image-cache/compare/v2.9.0...v3.1.0)

#### Improvements

- chore: updated dev and example packages ([b534f4f](https://github.com/georstat/react-native-image-cache/commit/b534f4f4717c1a3d0653a6f1ab013edc2d28b15f)).
- chore: removed husky, added lefthook ([f6d0745](https://github.com/georstat/react-native-image-cache/commit/f6d0745f27b5f5ea305946bd8303dc0141ff4a54)).
- chore: prettified, linted and a-b sorted some props ([708340d](https://github.com/georstat/react-native-image-cache/commit/708340dd328da559731cd1ca59f60b404fa223a3)).
- chore: renamed animated style variables for better readability ([3b7486b](https://github.com/georstat/react-native-image-cache/commit/3b7486bfa99ce8efd748ac9a8d790c8a3b0298f3)).
- chore: updated CHANGELOG.md ([2c019c1](https://github.com/georstat/react-native-image-cache/commit/2c019c1abc049c8680e488c8cba140859b710211)).
- chore: added 408, 500 & 503 error codes in the retry function ([e823000](https://github.com/georstat/react-native-image-cache/commit/e823000129d9e0b298e0463caa5768844f4d055f)).
- chore: use root on lefthook for example run ([66e04eb](https://github.com/georstat/react-native-image-cache/commit/66e04eb22ce5d3d96278c4f3ec2a1a6563721584)).
- chore: use `forEach` instead of `map` in prefetch function ([9ab5e91](https://github.com/georstat/react-native-image-cache/commit/9ab5e91ebfdeecf14372f6c87ded35f08e796107)).
- chore: yarn prettier now runs first on lefthook example run to avoid eslint issues afterwards ([5169b0e](https://github.com/georstat/react-native-image-cache/commit/5169b0e9a32fc2be648b9a6433ae0dac22825d9c)).

#### Documentations

- docs: added `imageStyle` prop and info about supporting local images ([23e2a70](https://github.com/georstat/react-native-image-cache/commit/23e2a70d01441b97d1bee9e1448d74e1123f4a28)).

## [v2.9.0](https://github.com/georstat/react-native-image-cache/compare/v2.8.0...v2.9.0) - 2023-08-10

#### Improvements

- chore: updated dev and example packages ([eba84f1](https://github.com/georstat/react-native-image-cache/commit/eba84f101269b150e1718fe4a1f13f8e50fc5d0e)).

#### Documentations

- docs: added supports new Arch in Features ([bd859ef](https://github.com/georstat/react-native-image-cache/commit/bd859ef2be9e1c4c603939925d0808021de04eda)).

## [v2.8.0](https://github.com/georstat/react-native-image-cache/compare/v2.7.0...v2.8.0) - 2023-08-07

#### Improvements

- chore: updated dev and example packages ([8633bbb](https://github.com/georstat/react-native-image-cache/commit/8633bbbdb3494dbe0fdc493f1a800617731034a1)).
- chore: import useRef instead of React.useRef ([d02ea03](https://github.com/georstat/react-native-image-cache/commit/d02ea03717374c5b76eb380fc37a0a95bfde66f8)).

## [v2.7.0](https://github.com/georstat/react-native-image-cache/compare/v2.6.0...v2.7.0) - 2023-07-31

#### Improvements

- chore: updated CHANGELOG.md ([4b01561](https://github.com/georstat/react-native-image-cache/commit/4b015611e5c2f0335dcd9cc572ce57aae85d09e8)).

#### Fixes

- fix: allow `0` as an animation duration on both source and thumbnail ([071941c](https://github.com/georstat/react-native-image-cache/commit/071941c850d0510f574d3d0b9fc7dfe0136d39da)).

## [v2.6.0](https://github.com/georstat/react-native-image-cache/compare/v2.5.0...v2.6.0) - 2023-07-31

#### Improvements

- chore: updated example packages, react-native@0.72.3 ([6ca18a8](https://github.com/georstat/react-native-image-cache/commit/6ca18a829b4190f8d58870635129a2bd868200c0)).
- chore: updated dev packages, react-native@0.72.3 ([6a5dd6d](https://github.com/georstat/react-native-image-cache/commit/6a5dd6d8f442cfa4195c66d193847caaa6363181)).
- chore: eslint fixes ([f25cbd5](https://github.com/georstat/react-native-image-cache/commit/f25cbd5ae9a701af11d6eb792a0ffeac01ec57c5)).

## [v2.5.0](https://github.com/georstat/react-native-image-cache/compare/v2.4.0...v2.5.0) - 2023-06-12

#### Improvements

- chore: updated dev end example packages ([d2359c1](https://github.com/georstat/react-native-image-cache/commit/d2359c146e9065993141470870d81f558b92ff41)).

## [v2.4.0](https://github.com/georstat/react-native-image-cache/compare/v2.3.0...v2.4.0) - 2023-05-12

#### Improvements

- chore: updated example to react-native@0.71.8 and dev packages ([c512549](https://github.com/georstat/react-native-image-cache/commit/c512549c46026012db3fc2cf60dcad2969838654)).
- chore: prettified files ([3288507](https://github.com/georstat/react-native-image-cache/commit/3288507f464a30a8e9f1a4a7364aae88e5fb4007)).
- chore: added nullish check on sourceAnimationDuration from global config ([f75ce87](https://github.com/georstat/react-native-image-cache/commit/f75ce870aa340ba31434d0d9986ca17af7c0439f)).
- chore: added tintColor prop on source image ([8c91921](https://github.com/georstat/react-native-image-cache/commit/8c9192124d0113aafd6f83ded8a6be8e2b710021)).
- chore: added missing / on bugs.url on package.json ([52cd27e](https://github.com/georstat/react-native-image-cache/commit/52cd27e6e05b0701cfc9f8eaa2598e8ce10ce17c)).
- chore: added missing comma on package.json ([457fcff](https://github.com/georstat/react-native-image-cache/commit/457fcfff73e1998aa0b03e90a5547e1eee9ee471)).
- chore: updated eslint rules ([232564d](https://github.com/georstat/react-native-image-cache/commit/232564d5d9900539206ef7979cbca7f787a9c6ef)).

#### Fixes

- fix: move reanimated to dev dependencies ([4066952](https://github.com/georstat/react-native-image-cache/commit/4066952f139fd4a5f8e8e3e76eb32f450a9008c7)).
- fix: use options for thumbnails & support token refresh

- Use options for thumbnails as well:
Options property are used only for source image. So, if you have any required header or option to get thumbnail images from whatever backend you're using, it'll silently fail.
- Support refresh on token update

src/CacheManager.ts
// @ts-ignore
import SHA1 from 'crypto-js/sha1';
import { FileStat, FileSystem } from 'react-native-file-access';

import { Config, DownloadOptions } from './types';
import defaultConfiguration from './defaultConfiguration';

async function retry(
  fn: () => any,
  retriesLeft = CacheManager.config.maxRetries || 0,
  interval = CacheManager.config.retryDelay
): Promise<any> {
  try {
    /* for some reason FileSystem.fetch won't throw error if image is not found
     * we need to catch the errors locally
     */
    const request = await fn();

    switch (request.status) {
      case 404:
        throw new Error(request.status);
      case 401:
        throw new Error(request.status);
      case 408:
        throw new Error(request.status);
      case 500:
        throw new Error(request.status);
      case 503:
        throw new Error(request.status);
      default:
        return request;
    }
  } catch (error: any) {
    /* abort early if the image is not found or
     * the access is not authorized
     */
    if (
      error.message === '404' ||
      error.message === '401' ||
      error.message === '500' ||
      error.message === '503' ||
      error.message === '408'
    ) {
      throw new Error(error);
    }
    /* FileSystem.fetch throws error if device is offline/temp internet loss with message "Host unreachable" or "Unable to resolve host"
     * so keep trying
     */
    if (retriesLeft === 0) {
      throw new Error(`Maximum retries exceeded: ${error.message}`);
    }
    await new Promise(resolve => setTimeout(resolve, interval));
    return retry(fn, retriesLeft - 1, interval);
  }
}

export class CacheEntry {
  source: string;

  options: DownloadOptions | undefined;

  downloadPromise: Promise<string | undefined> | undefined;

  pathResolved = false;

  noCache: boolean | undefined = false;

  maxAge: number | undefined;

  constructor(
    source: string,
    options: DownloadOptions | undefined,
    noCache?: boolean,
    maxAge?: number
  ) {
    this.noCache = noCache;
    this.options = options;
    this.source = source;
    this.maxAge = maxAge;
  }

  async getPath(): Promise<string | undefined> {
    const { source, maxAge, noCache } = this;
    const { exists, path } = await getCacheEntry(source, maxAge);

    if (exists && !noCache) {
      return path;
    }

    if (!this.downloadPromise) {
      this.pathResolved = false;
      this.downloadPromise = this.download(path);
    }

    if (this.downloadPromise && this.pathResolved) {
      this.pathResolved = false;
      this.downloadPromise = this.download(path);
    }
    return this.downloadPromise;
  }

  private async download(path: string): Promise<string | undefined> {
    const { source, options, noCache } = this;
    /* if noCache is true then return the source uri without caching it */
    if (noCache) {
      return source;
    }

    if (source != null) {
      try {
        const result = await retry(() =>
          FileSystem.fetch(source, {
            path,
            ...options,
          })
        );
        /* If the image download failed, we don't cache anything */
        if (result && result.status !== 200) {
          this.downloadPromise = undefined;
          return undefined;
        }
      } catch (e) {
        if (__DEV__) {
          console.log(
            `FileSystem.fetch has some trouble, error: ${
              e instanceof Error ? e.message : 'unknown'
            }`
          );
        }
        this.downloadPromise = undefined;
        return undefined;
      }

      if (CacheManager.config.cacheLimit) {
        await CacheManager.pruneCache();
      }
      this.pathResolved = true;
      return path;
    }
    return source;
  }
}

export default class CacheManager {
  static defaultConfig: Config = defaultConfiguration;

  static config: Config;

  get config() {
    return CacheManager.defaultConfig;
  }

  set config(newConfig) {
    CacheManager.defaultConfig = newConfig;
  }

  static entries: { [uri: string]: CacheEntry } = {};

  static get(
    source: string,
    options: DownloadOptions | undefined,
    noCache?: boolean,
    maxAge?: number
  ): CacheEntry {
    if (
      !CacheManager.entries[source] ||
      CacheManager.entries[source].options?.headers?.Authorization !==
        options?.headers?.Authorization
    ) {
      CacheManager.entries[source] = new CacheEntry(
        source,
        options,
        noCache,
        maxAge
      );
      return CacheManager.entries[source];
    }
    return CacheManager.entries[source];
  }

  static async clearCache(): Promise<void> {
    if (await FileSystem.exists(CacheManager.config.baseDir)) {
      const files = await FileSystem.ls(CacheManager.config.baseDir);
      for (const file of files) {
        try {
          await FileSystem.unlink(`${CacheManager.config.baseDir}${file}`);
        } catch (e) {
          if (__DEV__) {
            console.log(`error while clearing images cache, error: ${e}`);
          }
        }
      }
    }
  }

  static async removeCacheEntry(entry: string): Promise<void> {
    try {
      const file = await getCacheEntry(entry);
      const { path } = file;
      await FileSystem.unlink(path);
    } catch (e) {
      throw new Error('error while clearing image from cache');
    }
  }

  static async getCacheSize(): Promise<number> {
    const result = await FileSystem.stat(CacheManager.config.baseDir);
    if (!result) {
      throw new Error(`${CacheManager.config.baseDir} not found`);
    }
    return result.size;
  }

  static async isImageCached(entry: string): Promise<boolean> {
    try {
      const file = await getCacheEntry(entry);
      const { exists } = file;
      return exists;
    } catch (e) {
      throw new Error('Error while checking if image already exists on cache');
    }
  }

  static prefetch(source: string | string[], options?: DownloadOptions): void {
    if (typeof source === 'string') {
      CacheManager.get(source, options).getPath();
    } else {
      source.forEach(image => {
        CacheManager.get(image, options).getPath();
      });
    }
  }

  static async prefetchBlob(
    source: string,
    options?: DownloadOptions
  ): Promise<string | undefined> {
    const path = await CacheManager.get(source, options).getPath();
    if (path) {
      const blob = await FileSystem.readFile(path, 'base64');
      return blob;
    }
    return undefined;
  }

  static async pruneCache() {
    /* If cache directory does not exist yet there's no need for pruning. */
    if (!(await CacheManager.getCacheSize())) {
      return;
    }

    const files = await FileSystem.statDir(CacheManager.config.baseDir);

    files.sort((a: FileStat, b: FileStat) => {
      return a.lastModified - b.lastModified;
    });

    const currentCacheSize = files.reduce((cacheSize, file: FileStat) => {
      return cacheSize + file.size;
    }, 0);

    if (currentCacheSize > CacheManager.config.cacheLimit) {
      let overflowSize = currentCacheSize - CacheManager.config.cacheLimit;

      while (overflowSize > 0 && files.length) {
        const file = files.shift();
        if (file) {
          if (await FileSystem.exists(file.path)) {
            overflowSize = overflowSize - file.size;
            await FileSystem.unlink(file.path).catch(e => {
              if (__DEV__) {
                console.log(e);
              }
            });
          }
        }
      }
    }
  }
}

const getCacheEntry = async (
  cacheKey: string,
  maxAge?: number | undefined
): Promise<{ exists: boolean; path: string }> => {
  let newCacheKey = cacheKey;
  if (CacheManager.config.getCustomCacheKey) {
    newCacheKey = CacheManager.config.getCustomCacheKey(cacheKey);
  }
  const filename = cacheKey.substring(
    cacheKey.lastIndexOf('/'),
    cacheKey.indexOf('?') === -1 ? cacheKey.length : cacheKey.indexOf('?')
  );
  const ext =
    filename.indexOf('.') === -1
      ? '.jpg'
      : filename.substring(filename.lastIndexOf('.'));
  const sha = SHA1(newCacheKey);
  const path = `${CacheManager.config.baseDir}${sha}${ext}`;

  // TODO: maybe we don't have to do this every time
  try {
    await FileSystem.mkdir(CacheManager.config.baseDir);
  } catch (e) {
    /* do nothing */
  }
  const exists = await FileSystem.exists(path);

  if (maxAge && exists) {
    const { lastModified } = await FileSystem.stat(path);
    const ageInHours = Math.floor(Date.now() - lastModified) / 1000 / 3600;
    if (maxAge < ageInHours) {
      await FileSystem.unlink(path);
      return { exists: false, path };
    }
  }
  return { exists, path };
};
